### PR TITLE
feat(dialog-service): refactor consolidate open methods in one

### PR DESCRIPTION
**Pull request by StrahilKazlachev**
Starting point for #224, #223 
Will work on the tests after we clear the new open API.

### CODE REVIEW

**Comment by StrahilKazlachev** [comment]
@EisenbergEffect @PWKad What do you think about the `.open` method API? Is returning a structure like `{ openResult: Promise<OpenDialogResult>, closeResult: Promise<CloseDialogResult> }` too unfriendly? Do you prefer an API where you pass a setting to control the output:
```
open(settings: DialogSettings & { yieldController: false | undefined }): Promise<CloseDialogResult>;
open(settings: DialogSettings & { yieldController: true }): Promise<OpenDialogResult>;
```
Or you have something else?

**Comment by PWKad** [comment]
I was trying to find where I had mentioned this before but combing back through issues and gitter I couldn't.  I prefer the second that you had brought up before.

So the usage would be something like - 

```
export class MyViewModel {
  submit(){
    this.dialogService.open({ viewModel: Prompt, model: 'Good or Bad?'}).then(response => {
      if (!response.wasCancelled) { console.log('good'); }
    });
  }
}
```
And with the second flag
```
export class MyViewModel {
  submit(){
    this.dialogService.open({ viewModel: Prompt, model: 'Good or Bad?', yieldController: true }).then(controller => {
      controller.result.then(response => {
        if (!response.wasCancelled) { console.log('good'); }
      }
    });
  }
}
```
Right?

**Comment by StrahilKazlachev** [comment]
@PWKad On gitter, but I was left with the impression that you preferred the other solution, good thing I asked again 😅 For this signature though we will need #217 to properly type it, at least I don't see how to do the overloads with the current generator.
I'll write back more questions after pushing the changes.

**Comment by StrahilKazlachev** [comment]
@PWKad Updated the `.open()` API as discussed + tests. A couple of questions:
* naming
 * new interfaces: `OpenDialogResult`, `CloseDialogResult` - comments?
 * `OpenDialogResult#closeResult` or just `result`(too much **result** 😕 ). Will it be confusing/misleading if we name it `output` also(`DialogResult`/`CloseDialogResult`)?
* can you comment on #223 because I've designed `OpenDialogResult`(`#wasCancelled`) so that `.canActivate() => false` is considered a cancellation.

File: src/dialog-service.js [hunk 1]
```diff
@@ -34,32 +34,54 @@ export class DialogService {
...
+
+    closeResult.then(() => {
       _removeController(this, dialogController);
     }, () => {
       _removeController(this, dialogController);
     });
-    return _openDialog(this, childContainer, dialogController)
-      .then(() => dialogController);
+
+    let openResult = _getViewModel(this.container, this.compositionEngine, childContainer, dialogController).then((instruction) => {
```
Comment (PWKad) [inline src/dialog-service.js]: I think there should be a task to somehow reduce the complexity in here a bit later down the road but looks good for now!

**Comment by PWKad** [comment]
@StrahilKazlachev anything we need to answer before merging this?

**Comment by StrahilKazlachev** [comment]
@PWKad Will add some more tests, update readme and sample/example.

File: test/unit/dialog-renderer.spec.js [hunk 2]
```diff
@@ -5,7 +5,7 @@ import {dialogOptions} from '../../src/dialog-options';
 import {configure} from '../../src/aurelia-dialog';
 import {TestElement} from '../fixtures/test-element';
 
-describe('the Dialog Renderer', () => {
+describe('the Dialog Renderer', function () {
```
Comment (PWKad) [inline test/unit/dialog-renderer.spec.js]: Just curious if there was any reason to change these over?

Comment (StrahilKazlachev) [inline test/unit/dialog-renderer.spec.js]: `describe` and `it` set the context - `this`, using a lambda prevents it. If someone tries to use `this` without noticing a lambda has been used, he(me :sweat_smile:) can loose quite some time.

**Comment by PWKad** [comment]
Going to merge this first thing in the morning, please let me know if there is an order to merge in.

**Comment by StrahilKazlachev** [comment]
@PWKad no order for this one.

**Pull request merged**